Add Voxelab Aquila Support, additional HC32 details, ProUI fixes.

classicrocker883 opened the pull request:
<!--

Submitting a Pull Request

- Please fill out all sections of this form. You can delete the helpful comments.
- Pull Requests without clear information will take longer and may even be rejected.
- We get a high volume of submissions so please be patient during review.

-->

### Description
Since HC32 support was added, I would like to update it's details, as well as add `CR10_STOCKDISPLAY`/(12864 monochrome LCD) support for it.

Also, this adds Voxelab Aquila support for their other mainboard Chips. such as GD32 and N32 versions. labeled as `GD32F103RC_voxelab_maple` and `N32G455RE_voxelab_maple` respectively. they both use `BOARD_VOXELAB_AQUILA` which is totally seperate from the HC32 type, and should be treated as such.

Note: N32G455RE chip may technically be a 512K, however it is not capable of flashing anything more than 256K. that has been noted in **stm32f1-maple.ini**.

for some reason when compiling with `DWIN_LCD_PROUI` with the recent HC32 Aquila, some warnings show up. This aims to fix such warnings:

```r
proui\menus.cpp: In function 'void onDrawMenuItem()'
warning: the address of 'MenuItem::caption' will never be NULL

proui\dwin_lcd.cpp: In function 'void dwinWriteToMem()'
warning: this 'for' clause does not guard this statement, but the latter is misleadingly indented as if it were guarded by the 'for'

proui\bedlevel_tools.cpp: In static member function 'static void BedLevelTools::drawBedMesh()'
warning: narrowing conversion of ... 'xy_uint_t offset' from 'int' to 'short unsigned int'

HAL\STM32F1\MarlinSerial.cpp:31:29: warning: always_inline function might not be inlinable
'static __always_inline void my_usart_irq()'
```

---
edit:  
**dwinui.h** / **dwinui.cpp** changed `void drawInt ( ... int32_t value)` from **int32_t** to **long** as intended with the **inline**'s below. also this mitigates warning which pops up using some boards.

Having `BTT_SKR_MINI_E3_V3_0_1` with (at least) `default_envs = STM32F401RC_btt` results in failed compile and this error:  
'USART3' was not declared in this scope; - STM32/MarlinSerial.cpp - line 52

to fix this, **MB** SKR Mini E3 V3.0.1 was added to its siblings in **inc/Conditionals_adv.h** line ~1260  
```y
  #if MB(BTT_SKR_MINI_E3_V1_0, BTT_SKR_MINI_E3_V1_2, BTT_SKR_MINI_E3_V2_0, BTT_SKR_MINI_E3_V3_0, BTT_SKR_MINI_E3_V3_0_1, BTT_SKR_E3_TURBO, BTT_OCTOPUS_V1_1)
```


<!--

Clearly describe the submitted changes with lots of details. Include images where helpful. Initial reviewers may not be familiar with the subject, so be as thorough as possible. You can use MarkDown syntax to improve readability with bullet lists, code blocks, and so on. PREVIEW and fix up formatting before submitting.

-->

### Requirements

<!-- Does this PR require a specific board, LCD, etc.? -->

### Benefits

<!-- What does this PR fix or improve? -->

### Configurations
`DWIN_LCD_PROUI`
`CR10_STOCKDISPLAY`
<!-- Attach Configurations ZIP and any other files needed to test this PR. -->

### Related Issues
~~Please review **Marlin\src\HAL\STM32F1\MarlinSerial.cpp**~~  
~~and `static __always_inline void my_usart_irq()`~~

~~I know that the follow code may not be the correct way of doing things, other than reverting back to an earlier commit before removing `inline`.~~

~~however this code does work nonetheless~~
```y
static
#if MB(VOXELAB_AQUILA)
inline
#endif
__always_inline void my_usart_irq(
...
```
<!-- Does this PR fix a bug or fulfill a Feature Request? Link related Issues here. -->

classicrocker883 committed:
Fix menu.h _draw, fstr -> ftpl

classicrocker883 committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into bugfix-2.1.x-Nov4

classicrocker883 committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into bugfix-2.1.x-Nov4

classicrocker883 committed:
take care of warnings
proui\menus.cpp: "the address of 'MenuItem::caption' will never be NULL"
proui\dwin_lcd.cpp: "this 'for' clause does not guard..." In function 'dwinWriteToMem()'
proui\bedlevel_tools.cpp: narrowing conversion of ... 'offset' in function 'drawBedMesh()'
HAL\STM32F1\MarlinSerial.cpp: always_inline function might not be inlinable in function 'my_usart_irq()'
add `NO_MAPLE_WARNING` to inc/Warnings.cpp

classicrocker883 committed:
Add Voxelab Aquila support
support for GD32 and N32 mainboard versions.
add support for `CR10_STOCKDISPLAY` (12864 monochrome LCD) in HC32
update HC32 details

classicrocker883 committed:
DWIN displays work better w/ ENCODER_WAIT_MS 10

thinkyhead commented on Marlin/src/HAL/STM32F1/MarlinSerial.cpp:
```diff
@@ -28,7 +28,11 @@
 // Changed to handle Emergency Parser
-static __always_inline void my_usart_irq(ring_buffer *rb, ring_buffer *wb, usart_reg_map *regs, MSerialT &serial) {
+static
+#if MB(VOXELAB_AQUILA)
```
Instead of directly checking the MB here, define a flag (e.g., "`STATIC_INLINE_USART_IRQ`") in its pins file. It will then be included by `MarlinConfig.h`. `_TERN(STATIC_INLINE_USART_IRQ, inline)`.

Or, is it that `__always_inline` is defined in such a way that the extra `inline` is required everywhere for a certain STM32F1? If that is the case `__always_inline` could be patched in one of the `HAL/STM32F1/inc/Conditionals_*.h` files.

thinkyhead commented on Marlin/src/lcd/e3v2/proui/menus.cpp:
```diff
@@ -107,7 +107,7 @@ void onDrawMenuItem(MenuItem* menuitem, int8_t line) {
   if (menuitem->frameid)
     dwinFrameAreaCopy(menuitem->frameid, menuitem->frame.left, menuitem->frame.top, menuitem->frame.right, menuitem->frame.bottom, LBLX, MBASE(line));
-  else if (menuitem->caption)
+  else
```
Is it true that `menuitem` always has a `caption`?

classicrocker883 replied:
I think so. almost every instance uses one, this is for using Chinese characters, right? so if it doesnt have a "frameid" then it uses a caption, no?

classicrocker883 replied:
if you really wanted to what if you had it something like
```y
else if (strlen(menuitem->caption))
```
so that if `caption` has any value, it writes a caption. because what happens is for Chinese, `frameid` is true (>0), anytime its not chinese, `frameid` is always 0...false

edit:
i made that change and on the test it was successful, no warning. either or solution (without anything after `else`) works

thinkyhead commented on Marlin/src/pins/hc32f4/pins_AQUILA_101.h:
```diff
@@ -22,7 +22,7 @@
 
 //
-// Voxelab Aquila V1.0.1 and V1.0.2 (HC32F460) as found in the Voxelab Aquila X2
+// Voxelab Aquila V1.0.1/V1.0.2/V1.0.3 (HC32F460) as found in the Voxelab Aquila X2 and C2
```
If there are links to this board's specs or schematics online, this is the best place to include those.

classicrocker883 replied:
hm.. well, best I can give you is what I have found. 


<p><a target="_blank" rel="noopener noreferrer nofollow"><img align="center" src="https://preview.redd.it/262qxpgw1y961.jpg?width=960&crop=smart&auto=webp&s=6c308c2180093eacd7ebbf908759f88b6b53a251" width="480" height="360" style="max-width: 100%;"></a></p>
[an early official manual picture of the board]  

notice the chip is STM32 _RE_, the chip shortage came, they went right into GD32, and so on.

<br>
<br>
<p><a target="_blank" rel="noopener noreferrer nofollow"><img align="center" src="https://user-images.githubusercontent.com/6733752/182582428-287ae7da-9bd1-4079-b4c3-f399be06a114.png" width="480" height="360" style="max-width: 100%;"></a></p>
[this is an HC32 on an Aquila C2]  

I believe the **C2** only uses the V1.0.0 - HC32

<br>
<br>
<p><a target="_blank" rel="noopener noreferrer nofollow"><img align="center" src="https://frankfurt.apollo.olxcdn.com/v1/files/zpaiahaj1dl12-RO/image;s=1000x700" width="500" height="350" style="max-width: 100%;"></a></p>
[here is V1.0.2 - HC32]  

I have a V1.0.3 - HC32 myself.

classicrocker883 replied:
the most common we have is GD32 and N32 chip on a V1.0.1 board.  
cant say what differences are between versions. however one clear difference is even though the first V1.0.0 with STM32 has a BL touch port like all the rest of them, except the Aquila **C2**, which uses the same HC32 but even though the board says V1.0.0 it _doesn't_ have a BL touch port.

thinkyhead commented on ini/stm32f1-maple.ini:
```diff
@@ -142,6 +142,33 @@ extends              = env:STM32F103RE_creality_maple
+upload_protocol      = jlink
+
+[env:N32G455RE_voxelab_maple]
+extends              = env:STM32F103RC_maple
```
Is this RE or RC? If it's RE then it should have 512k, and so on….

thinkyhead replied:
Ah, I see… "cannot be flashed over 256K"

So what's up with these chips that they call themselves "RE" but then only have 256K? That's befuddling.

classicrocker883 replied:
> Ah, I see… "cannot be flashed over 256K"
> 
> So what's up with these chips that they call themselves "RE" but then only have 256K? That's befuddling.

I dont believe its the actual chip, but Voxelab using a bootloader with their own flavor. basically hard coding the limit, probably used the same one on the other variants.

this issue https://github.com/classicrocker883/MRiscoCProUI/issues/78 goes into alittle more detail about it  
as well as this disscusion here https://github.com/classicrocker883/MRiscoCProUI/discussions/86

thinkyhead replied:
We'll have to add a note in the INI explaining the limitation. We can also add an env: entry specifying upload by ST-Link (i.e., generic ICSP programmer) for a real 512K (>256K) build that uses `creality256k.ld`. There's no #warning currently for a build that "requires ICSP" so we can also add that.

thinkyhead commented on Marlin/src/lcd/e3v2/proui/bedlevel_tools.cpp:
```diff
@@ -247,7 +247,8 @@ bool BedLevelTools::meshValidate() {
       if (viewer_print_value) {
-        xy_uint_t offset { 0, cell_height_px / 2 - fs };
+        xy_uint_t offset;
+        offset.y = cell_height_px / 2 - fs;
```
This will be patched (differently) in another PR later today.

thinkyhead committed:
initial cleanup

thinkyhead committed:
Closer match to STM32F1

thinkyhead commented:
I split the doubled-up `HAL.cpp` file into two separate files and brought over some updates from the regular `HAL.cpp`. It will now be easier to compare the two HAL implementations and consolidate their common elements. It looks like `USE_USB_COMPOSITE` might be obsolete. It's no longer referenced in the main HAL for `__STM32F1__`. So, these are the parts of `HAL.cpp` that maybe should be brought over to `HAL_N32.cpp`:

<img width="812" alt="image" src="https://github.com/MarlinFirmware/Marlin/assets/698003/86d224ce-cfbb-4463-af92-e86d11f708f5">

thinkyhead committed:
consolidate

thinkyhead committed:
Perhaps (?)

thinkyhead committed:
headers besides HAL.h get Cond_post

thinkyhead committed:
cleanup

thinkyhead committed:
translate

thinkyhead committed:
one less

classicrocker883 committed:
Update pins_AQUILA_101.h

thinkyhead committed:
fix adc

thinkyhead committed:
redundant

thinkyhead committed:
Already runs custom_board.py

thinkyhead committed:
allow 512k build, but warn

thinkyhead committed:
just wrap ADC init

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Nov4

thinkyhead committed:
✅ Temporary CI Tests for 2.1.x

thinkyhead committed:
SAMD51 ADCs

ellensp committed:
🧑‍💻 More num-to-string digits / precisions (#26343)

Co-authored-by: Scott Lahteine <thinkyhead@users.noreply.github.com>

classicrocker883 committed:
Update hc32.ini

classicrocker883 committed:
Update boards.h

classicrocker883 committed:
fix drawInt to "long value"
int32_t was meant to be "long" as we see with the "inline"
short and long can vary between different compilers and systems.
so here we just use long as intended. ~~(int64_t = long long)~~

thinkyhead committed:
[cron] Bump distribution date (2023-12-02)

classicrocker883 committed:
fix error for BTT_SKR_MINI_E3_V3_0_1
'USART3' was not declared in this scope; - STM32/MarlinSerial.cpp - line 52
overflow in conversion from 'PinName' to 'pin_t' - STM32/MarlinSPI.h - line 55

thinkyhead committed:
keep old names

thinkyhead committed:
shorter

thinkyhead commented on Marlin/src/inc/Conditionals_adv.h:
```diff
@@ -1256,7 +1256,7 @@
  */
 #if LCD_IS_SERIAL_HOST && !defined(LCD_SERIAL_PORT)
-  #if MB(BTT_SKR_MINI_E3_V1_0, BTT_SKR_MINI_E3_V1_2, BTT_SKR_MINI_E3_V2_0, BTT_SKR_MINI_E3_V3_0, BTT_SKR_E3_TURBO, BTT_OCTOPUS_V1_1, AQUILA_V101)
+  #if MB(BTT_SKR_MINI_E3_V1_0, BTT_SKR_MINI_E3_V1_2, BTT_SKR_MINI_E3_V2_0, BTT_SKR_MINI_E3_V3_0, BTT_SKR_MINI_E3_V3_0_1, BTT_SKR_E3_TURBO, BTT_OCTOPUS_V1_1)
```
What serial port does the `AQUILA_V101` prefer for LCD usage?

classicrocker883 replied:
oops i didnt mean to remove that! hang on let me fix it

thinkyhead commented on Marlin/src/lcd/e3v2/proui/dwinui.cpp:
```diff
@@ -181,7 +181,7 @@ void DWINUI::drawString(uint16_t color, const char * const string, uint16_t rlim
 //  x/y: Upper-left coordinate
 //  value: Integer value
-void DWINUI::drawInt(uint8_t bShow, bool signedMode, fontid_t fid, uint16_t color, uint16_t bColor, uint8_t iNum, uint16_t x, uint16_t y, int32_t value) {
+void DWINUI::drawInt(uint8_t bShow, bool signedMode, fontid_t fid, uint16_t color, uint16_t bColor, uint8_t iNum, uint16_t x, uint16_t y, long value) {
```
Is a `long` of undetermined size required here? By using types with fixed bit lengths we avoid that ambiguity and never have to deal with any 64-bit `long` type, now or in future.

classicrocker883 replied:
hm. well... I suppose we can guarantee 32-bit with `int32_t`. I did a quick look through and looks to me most "value"'s of the function are calling for 16. I figured since the rest of the inline are cast to `long` they all should be. what do you think?

thinkyhead replied:
Since they all the `drawInt` methods take `long` –and `long` is 32-bits on both AVR and ARM– it should be fine.

classicrocker883 committed:
revert Conditionals_adv.h
put back Aquila_V101

classicrocker883 committed:
Merge branch 'bugfix-2.1.x-Nov4' of https://github.com/classicrocker883/Marlin into bugfix-2.1.x-Nov4

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into pr/26470

classicrocker883 committed:
fix for STM32F401RC_btt
overflow in conversion from 'PinName' to 'pin_t' - STM32/MarlinSPI.h - line 55

classicrocker883 committed:
Merge branch 'bugfix-2.1.x-Nov4' of https://github.com/classicrocker883/Marlin into bugfix-2.1.x-Nov4

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26470

thinkyhead commented:
> Merge branch 'bugfix-2.1.x-Nov4' of https://github.com/classicrocker883/Marlin into bugfix-2.1.x-Nov4

You should be using `git merge upstream/bugfix-2.1.x` followed by `git push` if your intention is to stay in sync with `bugfix-2.1.x` updates.

thinkyhead commented on Marlin/src/HAL/STM32/HAL.h:
```diff
@@ -138,7 +138,7 @@
 typedef double isr_float_t;   // FPU ops are used for single-precision, so use double for ISRs.
 
-#if defined(STM32G0B1xx) || defined(STM32H7xx)
+#if defined(STM32G0B1xx) || defined(STM32H7xx) || defined(STM32F401xC)
```
Is this now an exhaustive list of all STM32 MCUs that require 32-bits for their pin numbers? I'm sure all processors have 32768 pins or fewer, but perhaps this 32-bits makes using the values faster due to ARM/RISC voodoo.

classicrocker883 replied:
that shouldnt be all STM32, only just those 3?

thinkyhead replied:
Yes, that is the burning question. Do some STM32 require a 16-bit type for GPIO while others need 32-bit? AFAIK pin numbers are just sequential integers, not port numbers and bit masks (but port numbers and bit masks could be a good scheme). The difference in type implies that these three are importing different platform code that expects 32-bit, but other STM32 are importing code that expects 16-bit. Since we import essentially the same platform code for all STM32 devices, it must be happening behind the scenes with the STM32 headers selectively including.

I'll have to Google around to see if there's some definitive answer about this. The STM32 variants all refer to a `PinName` type, so that's a clue where I can start looking.

thinkyhead replied:
`PinName` is just an untyped `enum` so I believe that makes its type the same size as `int`.

Where it all began…

The first STM32 HAL just used `uint8_t` for `pin_t`. Then it got bumped to 16 bits by #14523 (for STM32F407). Then in #23283 it got 32 bits for STM32G0B1xx, keeping 16 bits for others. Then #25387 added on STM32H7xx. With this PR adding STM32F401xC, it got me wondering.

STMicro's STM32 platform expects `uint32_t` for pin conversion functions like `digitalPinToPinName`. But `framework-arduino-hc32f46x` defines and uses `typedef int16_t gpio_pin_t` and `framework-arduinoststm32-maple` uses only 8 bits for functions like `gpio_set_pin`.

So for HC32 using Maple it may be more efficient to use 16 bit pin numbers. In general we want to find the best balance between size and speed. Different-sized types for pin numbers may or may not increase the size of the code and SRAM usage. But if a larger type results in faster (and potentially smaller) code then the tradeoff may be worthwhile.

Hmm…

thinkyhead replied:
Interestingly, Marlin code size for a HAL/STM32 build is no different whether `pin_t` is defined with 8, 16, or 32 bits. So it seems best to just go with 32-bits for the best parity with `platforms/ststm32`.

classicrocker883 replied:
it would be a good idea for a 2nd opinion. @shadow578 seems to know HC32 in and out. also remember HC32 is Cortex-M4 and should be treated separately even though it takes cue from STM32F1-Maple.

the fix I found for the warning "overflow in conversion from 'PinName' to 'pin_t'" happened by using default_envs = `STM32F401RC_btt`. I haven't try seeing if pops up for any other STM32F401RC platforms, like Ender-3 S1.

so we could see through the Tests/Checks logs if that kind of warning shows up. I mean if that warning isn't present (before this **PR** change in question)

I guess what you're wondering is would it benefit ALL the chips? well, if it is good for one, it should be good for all?

then that may take some actual trial and error, or real world testing. if theres something I can do on my end just let me know.

thinkyhead committed:
developer note

thinkyhead committed:
or, rather…

thinkyhead committed:
just in case

thinkyhead committed:
32 bit pin type

classicrocker883 commented:
I just wanted to quickly mention some **N32** chips are N32G45**2**RE. the difference I can only find between them is:

G45**5** - 12-bit ADC: 4/22-ch
G45**2** - 12-bit ADC: 2/16-ch

taken from here [N32G455](https://www.nationstech.com/N32G455/)
taken from here [N32G452](https://www.nationstech.com/N32G452/)

probably not significant enough to bother adding/changing comments provided?

classicrocker883 committed:
Update adc.h

POWERMON_VOLTS to POWERMON_VOLTAGE

classicrocker883 commented:
changes to the **HAL** for **N32** results in additional 152 bytes in Flash and 8 in RAM. I would think it would stay the same. any idea what would account for the differences? and is it for better or for worse??

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26470

thinkyhead committed:
Try FORCE_INLINE

thinkyhead commented:
> changes to the HAL for N32 results in additional 152 bytes in Flash and 8 in RAM. I would think it would stay the same. any idea what would account for the differences? and is it for better or for worse??

Considering that Maple is deprecated for all MCUs except those that require it, a few bytes of Flash and SRAM are not problematic in the grand scheme. Meanwhile, the only difference in the build for the `STM32F103VE_longer_maple` (for example) is +16 bytes of Flash.
